### PR TITLE
8296101: nmethod::is_unloading result unstable with concurrent unloading

### PR DESCRIPTION
**Pull request by fisk**
If an nmethod is not called during a concurrent full GC, then after marking has terminated, multiple threads can call is_unloading. If at the same time, the nmethod is made not_entrant, then we run into a source of instability in the is_cold() calculation used when computing is_unloading. There we check if the nmethod is_not_entrant(), which some concurrent observers will think is true, while others think it's false.
The current code that sets the is_unloading_state in is_unloading() assumes that the computed state is the same across all observers. However, that is no longer true.

I propose to set the is_unloading_state with a CAS instead of plain store. Then, as is_unloading() is computed before making nmethods not_entrant, we can guarantee that all concurrent readers of is_unloading in this scenario will return false in the current unloading cycle, instead of racingly returning either false or true. One thread wins, and it will say false, and the other threads will compute conflicting results, but end up agreeing after the CAS, that they should all return false.

Tested with mach5 tier1-7. Also tried replacing the is_not_entrant() ingredient in is_cold with os::random() to simulate the instability source. Without my fix RunThese crashes almost immediately, and with my fix it doesn't crash.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296101](https://bugs.openjdk.org/browse/JDK-8296101): nmethod::is_unloading result unstable with concurrent unloading

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10926/head:pull/10926` \
`$ git checkout pull/10926`

Update a local copy of the PR: \
`$ git checkout pull/10926` \
`$ git pull https://git.openjdk.org/jdk pull/10926/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10926`

View PR using the GUI difftool: \
`$ git pr show -t 10926`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10926.diff">https://git.openjdk.org/jdk/pull/10926.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/code/nmethod.cpp [hunk 1]
```diff
@@ -1655,11 +1655,22 @@ bool nmethod::is_unloading() {
...
+  // is_unloading or not.
+  if (Atomic::cmpxchg(&_is_unloading_state, state, new_state, memory_order_relaxed) == state) {
+    // Our view of the world won
+    return state_is_unloading;
+  }
+
+  // If we lost the race, we call is_unloading again to see what the winner
+  // determined. This can't recurse as the second time we call the state is
+  // guaranteed to be cached for the current unloading cycle.
+  return is_unloading();
```
Comment (kimbarrett) [inline src/hotspot/share/code/nmethod.cpp:1673]: FWIW, it seems like you could remove any question about recursion by doing
something like this:
```
uint8_t found_state = Atomic::cmpxchg(&_is_unloading_state, state, new_state, memory_order_relaxed);
uint8_t updated_state = (found_state == state) ? new_state : found_state;
return IsUnloadingState::is_unloading(updated_state);
```
Is that clearer (esp. with corresponding commentary)?  Maybe.

**Review by kimbarrett — APPROVED** [review]
Looks good as-is, but with a suggested alternate you can take or not.

File: src/hotspot/share/code/nmethod.cpp (hunk 1 above)
Comment (dean-long) [reply src/hotspot/share/code/nmethod.cpp:1673]: I think avoiding recursion is clearer.  How about:

```
if (found_state == state) {
  // First to change state, we win
  return state_is_unloading;
} else {
  // State already set, so use it
  return IsUnloadingState::is_unloading(found_state);
}

```

**Comment by fisk** [comment]
Thank you for the reviews, @kimbarrett and @dean-long ! I updated the PR with a variation without recursion, as requested.

**Review by kimbarrett — APPROVED** [review]
Recursion removal looks good.

**Comment by fisk** [comment]
Thanks for the reviews, @kimbarrett and @dean-long .

**Pull request closed without merging**